### ISSUE TITLE
MSMQ: When the queue name is too long, although there is a warning, an unusable queue is still created. 

### ISSUE THREAD

**Issue by indualagarsamy** [Particular/NServiceBus#4630]
**Steps to reproduce:**

1. Setup a queue name to be more than the accepted length of characters.  
I used: `NServiceBus.Core.Tests.MsmqQueueCreatorTests.ReceivingBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB`
2. When the `CreateQueueIfNecessary` in `MsmqQueueCreator.cs` is invoked, while it checks if the queue exists, and if its remote, if not, it invokes the Create method. But the create method will create the queue successfully, and setting the queue permissions will fail and we catch that exception and log a warning. 
3. But then we have an unusable queue that appears to be stuck and cannot be deleted. 

**My Proposal:**
- Log a warning, if the queue name length exceeds 63. As far as why 63, see:  http://geekswithblogs.net/Plumbersmate/archive/2010/09/13/how-long-msmq-queue-names-are-displayed-.-or-not.aspx
- Throw an exception, if the queue name length exceeds 124 (which includes the format name length), because otherwise, there is the potential to have an unusable queue. 124 is as documented here: 
https://msdn.microsoft.com/en-us/library/ms706083(v=VS.85).aspx

**Letting the function create the queue:**
- Unable to delete the queue. And the queues seem to be there even after restarting computer management. 
  <img width="1197" alt="screen shot 2017-04-19 at 8 35 46 pm" src="https://cloud.githubusercontent.com/assets/854553/25212582/0322b680-2540-11e7-8223-32d5f869630e.png">
- Causes QueueExplorer to act up as well. 

**How to delete the weird queue that cannot be deleted from the UI:**
Found the instructions from SO. - http://stackoverflow.com/questions/1561063/msmq-cannot-delete-or-purge-a-queue
- Stop MSMQ service
- In a text editor, inspect the files in C:\Windows\System32\msmq\storage\lqs and locate the file with the queue name and delete the file
- Restart MSMQ service.

**Comment by indualagarsamy** [comment]
@Particular/msmq-transport-maintainers - If you agree with the proposed fix, post a comment here. I am working on a PR. I discovered this when I was working on the PR for unit test cleanup.

**Comment by andreasohlund** [comment]
@ramonsmits @timbussmann I have a distinct memory of us leaving this as is for some backwards compatibility reason but I can't remeber the exact details?

**Comment by indualagarsamy** [comment]
@andreasohlund @ramonsmits @timbussmann - I don't think it's a good idea to land the user in a potentially unusable situation when creating new endpoints. The change I am talking about is when we create queues. I assume this gets called when a new endpoint is being setup? 
Already existing queues should not matter.
Code in question is:
https://github.com/Particular/NServiceBus/blob/develop/src/NServiceBus.Core/Transports/Msmq/MsmqQueueCreator.cs#L31

**Comment by indualagarsamy** [comment]
Updated the issue description with instructions on manually deleting the weird queue.

**Comment by indualagarsamy** [comment]
I think we can still do this in a backwards compatible way:

- Always log a warning if the queue length is suspect. See my previous link on the length based on John Breakwell's blog.
- Not throw (if its a valid queue length) but less than the number where SetPermissions blows up. 
- Throw if its not a valid queue length(Format name > 124) . The chances of such queues in Prod are less, super edge case is unless they went through some serious interop to change the buffer value as described in the link I provided. This will help new users who are naming their endpoints with super long names who might could potentially hit this condition? 

Thoughts?

**Comment by andreasohlund** [comment]
@indualagarsamy I raised a similar issue recently (that we closed as won't fix), can you eyeball it and see how it aligns with this?

https://github.com/Particular/NServiceBus/issues/4324

**Comment by indualagarsamy** [comment]
@andreasohlund - Thanks for the history on #4324. Essentially, it is the same issue, however,  I think the symptoms are more severe than what was listed in that issue, i.e.

>Users with a endpoint name resulting in a format name above 102 characters will see warnings in the log stating that the queues couldn't be checked and the queues won't have the requested permissions.

Its, not just the warnings about queue permissions, but potentially creating an unusable queue, which may not be displayed properly in both Computer Management tool and also cause problems for other tools like Queue Explorer and cannot be deleted using UI tools. The only way to delete such a queue is to resort to manual methods of going to disk and looking at all the queue files to identify and delete as I have described in the issue description.

Also, the difference between the issue you had raised and this one in when we should trigger a warning is the number of characters. I am not sure how the 102 character limit was reached. In my tests, a format name length of 114 worked for setting queue permissions. Is there any documentation on the validation length?

My proposal is use John Breakwell's best practice for the warning in the very least: "Best practice - stick to 63 characters for public or private queues and you won't go wrong." As this has implications on how the queues are displayed in PerformanceMonitor as well. 

I am glad that our users have not run into this. I do think that it will be an easy thing for us to check and caution our users if we think the length might be a cause for concern. And to throw an exception if we thing that the length might create an unusable queue. That would make us safe by default.  

Please ping me if you think we need to discuss this on a call.

**Comment by indualagarsamy** [comment]
The MMC Snap In has a maximum limit for the queue name to be 115 characters. However, a 115 character queue name when converted to the format name, which includes additional format characters, including the machine name can easily top the buffer size which is 124, hardcoded in the messaging .NET library. 

 In Version 5.x, when the format name exceeds 123 characters there will be an exception. This dates back to NServiceBus Version 4.0.0, when we added ACL permissions.
https://github.com/Particular/NServiceBus/commit/76089325ff7c57208e4a0794baf882750db2dbfa. ACL with a format name greater than 123 characters would have resulted in the endpoint not starting. 

So, in short, we can safely throw an exception if the endpoint name exceeds the format name limit. Backwards compatibility should not be a problem. Since we create the subqueues like, .timeoutdispatcher etc, we should check that our sub queues won't trigger this as well.

**Comment by dvdstelt** [comment]
Moved to https://github.com/Particular/NServiceBus.Transport.Msmq/issues/5.
